yt/yt/server/query_tracker/query_tracker_proxy.cpp

```diff
@@ -968,34 +1111,54 @@ TListQueriesResult TQueryTrackerProxy::ListQueries(
         }
     };
 
-    auto selectFinishedQueries = [=, this, this_ = MakeStrong(this)] () -> std::vector<TQuery> {
+    bool incomplete = false;
+    auto selectFinishedQueries = [=, this, this_ = MakeStrong(this), &incomplete] () -> std::vector<TQuery> {
         try {
             TStringBuilder admittedQueryIds;
+
             {
-                NQueryClient::TQueryBuilder builder;
-                TYsonString placeholderValues;
-                builder.SetSource(StateRoot_ + "/finished_queries_by_start_time");
-                addFilterConditions(builder, placeholderValues);
-                builder.AddSelectExpression("[query_id]");
-                TSelectRowsOptions options;
-                options.Timestamp = timestamp;
-                options.PlaceholderValues = placeholderValues;
-                auto query = builder.Build();
-                YT_LOG_DEBUG("Selecting finished queries by start time (Query: %v)", query);
-                auto selectResult = WaitFor(StateClient_->SelectRows(query, options))
-                    .ValueOrThrow();
+                std::vector<std::pair<TTimestamp, TQueryId>> results;
+                if (isSuperuser) {
```

Comment:
It would be nice to check here `if (isSuperuser && UserFilter.empty())` to improve performance when superuser asks about a specific user (themselves, for example)